File: problem133.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

"""
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        
        if not node:
            return None
        
        adjList = defaultdict(set)
        visited=set()
        stack = []
        stack.append(node)
        while(stack):
            current=stack.pop()
            adjList[current.val]=set()
            for neighbor in current.neighbors:
                adjList[current.val].add(neighbor.val)
                if neighbor.val not in visited:
                    visited.add(neighbor.val)
                    stack.append(neighbor)
        
        #print(adjList)
        result = {}
        for k in adjList.keys():
            result[k] = Node(k,[])
            
        for k in adjList.keys():
            result[k].neighbors = [result[n] for n in adjList[k]]
        
        return result[node.val]

File: problem133.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> neighbors;

    public Node() {
        val = 0;
        neighbors = new List<Node>();
    }

    public Node(int _val) {
        val = _val;
        neighbors = new List<Node>();
    }

    public Node(int _val, List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}

visited=[1,2]
nodes = {
1:2,
2:
}
                            [1]
                        /       \

*/

public class Solution {
    Dictionary<int, Node> nodes;
    HashSet<int> visited;
    public Node CloneGraph(Node node) {
        
        if(node == null)
            return null;
        
        nodes = new Dictionary<int, Node>();
        visited = new HashSet<int>();
        visited.Add(node.val);
        Dfs(node);
        
        //PrintGraph(nodes[node.val]);
        
        return nodes[node.val];
    }
    
    public void Dfs(Node node)
    {
        if(node==null)
            return;
        
        if(!nodes.ContainsKey(node.val))
            nodes[node.val] = new Node(node.val);
        
        foreach(Node n in node.neighbors)
        {
            if(!nodes.ContainsKey(n.val))
                nodes[n.val] = new Node(n.val);
            
            if(!nodes[node.val].neighbors.Contains(nodes[n.val]))
                nodes[node.val].neighbors.Add(nodes[n.val]);
            
            if(!nodes[n.val].neighbors.Contains(nodes[node.val]))
                nodes[n.val].neighbors.Add(nodes[node.val]);
            
            if(!visited.Contains(n.val))
            {
                visited.Add(n.val);
                Dfs(n);
            }
        }
        
        return;
    }
    
    void PrintGraph(Node node)
    {
        Queue<Node> queue=new Queue<Node>();
        HashSet<int> visited = new HashSet<int>();
        
        queue.Enqueue(node);
        visited.Add(node.val);
        
        while(queue.Count>0)
        {
            Node current = queue.Dequeue();
            foreach(Node neighbor in current.neighbors)
            {
                if(!visited.Contains(neighbor.val))
                {
                    visited.Add(neighbor.val);
                    queue.Enqueue(neighbor);
                }
            }
        }
        return;
    }
}

//time: O(N)
//space: O(N)